### PR TITLE
feat: replacement support for more autoreplace managers

### PR DESCRIPTION
**Pull request by t-kulmburg**
## Changes

This is based on a [PR](https://github.com/renovatebot/renovate/pull/13211) from @JamieMagee about adding support for replacements for autoreplace managers

## Context

After fixing building errors I added the name-replacement into the doAutoReplace function as they share most of their logic.
If you have any good ideas for better structure, I can refactor the code further.

I also crated a [gitlab demo-repo](https://gitlab.com/t.kulmburg/replacement-demo/-/blob/main/renovate.json) containing a dependency for all managers listed in the PR of @JamieMagee.
As can be seen in the [repo README](https://gitlab.com/t.kulmburg/replacement-demo/-/blob/main/README.md) and the [Renovate created merge requests](https://gitlab.com/t.kulmburg/replacement-demo/-/merge_requests), more then half of them are already working with my rather minimal changes.
I also added a brief description of the problems the other managers currently have.
To test the feature even further I added unit tests for Managers that worked in the demo-repo.

My question now would be how you would like to proceed with this topic in general.
Getting the feature to work for all supported managers will probably require a lot more work,
but supporting only the currently working ones will require a detailed documentation on the current status.

I am willing to spend some more time on this in the next days/weeks, so reviews would be appreciated.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/workers/repository/update/branch/auto-replace.spec.ts
Comment (viceice) [inline lib/workers/repository/update/branch/auto-replace.spec.ts:193]:
```suggestion
      upgrade = getConfig();
```

File: lib/workers/repository/update/branch/get-updated.spec.ts
Comment (viceice) [inline lib/workers/repository/update/branch/get-updated.spec.ts:513]:
```suggestion
      expect(res).toMatchObject({
```

Please read our best practices guide in `docs/development` directory

File: lib/workers/repository/update/branch/auto-replace.ts
Comment (viceice) [inline lib/workers/repository/update/branch/auto-replace.ts:131]: Please do this refactoring in new PR, so we can merge that big change first

File: lib/workers/repository/update/branch/auto-replace.spec.ts
Comment (viceice) [inline lib/workers/repository/update/branch/auto-replace.spec.ts:278]: shift a level up and reuse for all tests

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/workers/repository/update/branch/auto-replace.spec.ts
Comment (t-kulmburg) [reply lib/workers/repository/update/branch/auto-replace.spec.ts]: The config used in this file before my changes is also created via json.parse.
getConfig returns an AllConfig object as opposed to the BranchUpgradeConfig used in the doAutoReplace function.
Multiple properties used are not part of AllConfig.

File: lib/workers/repository/update/branch/get-updated.spec.ts
Comment (t-kulmburg) [reply lib/workers/repository/update/branch/get-updated.spec.ts]: done

File: lib/workers/repository/update/branch/auto-replace.spec.ts
Comment (t-kulmburg) [reply lib/workers/repository/update/branch/auto-replace.spec.ts]: done

File: lib/workers/repository/update/branch/auto-replace.spec.ts
Comment (viceice) [reply lib/workers/repository/update/branch/auto-replace.spec.ts]: see updated suggestions

File: lib/workers/repository/update/branch/auto-replace.spec.ts
Comment (viceice) [reply lib/workers/repository/update/branch/auto-replace.spec.ts]: please read our contribution guide[^1]

[^1]: https://github.com/renovatebot/renovate/blob/main/.github/contributing.md#resolve-review-comments-instead-of-commenting

**Comment by JamieMagee** [comment]
@t-kulmburg Thanks for taking this over. I am going to close #13211 in favour of your PR.

**Comment by t-kulmburg** [comment]
I just pushed a fix to enable replacing the same dependency with different other dependencies, based on the 'matchCurrentVersion' constraint.[ Example Mr](https://gitlab.com/t.kulmburg/replacement-demo/-/merge_requests/62/diffs).

I am now investigating and trying to fix the problem that replacement doesnt work for docker images with digest.

If you want any documentation/more tests etc. just let me know!

**Comment by t-kulmburg** [comment]
Another update:

 * Docker digests are now also updated correctly. [MergeRequest](https://gitlab.com/t.kulmburg/replacement-demo/-/merge_requests/30/diffs#6651ddff6eb82c840ced7c1dddee15c6e1913dd4)
 * Python version restrictions (pep440) are now ignored for replacements (as the replacement version could carry it's own restrictions), these now also work.  [MergeRequest](https://gitlab.com/t.kulmburg/replacement-demo/-/merge_requests/25/diffs?view=inline#15af4b3d30054104e85f65a444b1db188c5deefb)
 * Found an issue with kustomize, the replacement is not as expected --> will investigate tomorrow

There are several manager that would need additional fields, to be fully supported (gitlab-ci-include needs file, helmv3 would need URL etc., all documented in my demo [README](https://gitlab.com/t.kulmburg/replacement-demo/-/blob/main/README.md)).
My proposal is that I fix kustomize, so that there are no "bugs" only missing "features", that could be implemented at a later date.

I plan to mark my PR as "Ready for Review" as soon as I have that kustomize issue resolved,
or do you have anything else you would want me to do, before this feature is merge worthy?

**Comment by t-kulmburg** [comment]
Kustomize is actually replacing as expected, but the option to replace subdirectories is missing as they are not part of the 'depName': [MR](https://gitlab.com/netconomy/personal/t.kulmburg/renovate-test/-/merge_requests/26/diffs)

**Comment by chezsmithy** [comment]
@t-kulmburg is it possible to replace the docker registry and the image with this change?

**Comment by t-kulmburg** [comment]
> @t-kulmburg is it possible to replace the docker registry and the image with this change?

I assume you mean in a Dockerfile, yes it is possible.
I added two more examples to the demo repo: [MR-1](https://gitlab.com/t.kulmburg/replacement-demo/-/merge_requests/99/diffs),  [MR-2](https://gitlab.com/t.kulmburg/replacement-demo/-/merge_requests/65/diffs)

**Comment by chezsmithy** [comment]
@t-kulmburg you might need to request a new review from viceice and paging @JamieMagee for a review.

File: lib/config/utils.ts
Comment (viceice) [inline lib/config/utils.ts:23]: why this change? 🤔

File: lib/modules/datasource/index.ts
Comment (viceice) [inline lib/modules/datasource/index.ts:437]: why this change?

**Review by viceice — CHANGES_REQUESTED** [review]
i think this should be split into multiple PR.
- datasource changes
- worker changes
- auto-replace changes
- manager changes
- versioning changes

File: lib/config/utils.ts
Comment (t-kulmburg) [reply lib/config/utils.ts:23]: Without this I had the problem that my newName, newValue and finally depName changed from one string to an array of single characters.

File: lib/modules/datasource/index.ts
Comment (t-kulmburg) [reply lib/modules/datasource/index.ts:437]: this "packageName" is used to look up digests. After this change it is able to replace images with digests, as now the correct digest for the replacement is used. Without this the depName would be replaced, but the old digest would be used as the old name would be used in lookUp

**Comment by t-kulmburg** [comment]
> 

Created 3 new PRs:
- #17881
- #17882 
- #17883

should I close this PR now?

**Pull request closed without merging**
